### PR TITLE
added narrativewriter and made ui changes

### LINKED ISSUE #4935: Export Narrative to MS Word
As an ***analyst***, I wish to ***view a combined set of narratives, outside Debrief, while I am waiting for the track data to arrive & be processed***, in order to ***familiarise myself with what happened, and plan my analysis***

I expect **Apache POI-Word** is the best library to use for this task.

* [x] Add "Export to MS Word" button to Narrative Viewer in Full Debrief.
* [x] Re-use the existing "Export to WMF" icon, in `icons/16/ex_2word.png`
* [x] Create a `.docx` file. The filename will be the date-time of the first narrative entry, formatted as `yyyymmdd_hhmm`.
* [x] Put a single table in the document.  `Time` and `Entry` will always be present. Decide on whether to include `Source` and `Type` depending on if those columns are currently visible
* [x] Export all visible narrative entries (so use can use Outline view to decide what to include)
* [x] Popup file dialog to choose target directory.  The default should be the previous directory, if possible, otherwise the user's home folder.

_Just in case_ we ever want to support this in D-Lite, most functionality should be a in a non-UI class within `cmap.legacy` poss in new `MWC.Utilities.ReaderWriter.Word` package.  Maybe a `NarrativeWriter` class with this API:
```
    public void write(NarrativeWrapper narratives, boolean showSource, boolean showType, File targetDirectory);
```

### PR DESCRIPTION
**Pull request by ayeshare**
Fixes #4935
Added the narrativewriter class to export to word

### CODE REVIEW

**Comment by helenayele** [comment]
Okay is this still work in progress? I couldn't see the Export to Word button in narrative viewer .
[Image from Gyazo](https://gyazo.com/59767b320d701e4d3a009b8fbaeca995)

**Comment by ayeshare** [comment]
> See that I already exported the narrative entry before, and when I try to export again, it doesn't replace it.

@helenayele I tried this out and it is working.  The filename remains the same.  Try exporting without a source or type and see if it works.  Also, you could try checking the timestamp of the file.

**Comment by IanMayo** [comment]
Replies to @helenayele 

2. File overwrite.  
I'm ok with just overwriting the data.  The user can re-create the documents quickly & easily.

3. Filename
If we change what is visible, we're changing which data is to be exported.  Then it's ok to use a different filename.

**Review by helenayele — CHANGES_REQUESTED** [review]
1. Export visible narrative files to MS Document allows us to export blank file. And doing this throws Nullpointer exception. See this video 
[Image from Gyazo](https://gyazo.com/69aa22fd72cf4b16a6ae039741e09c33)
So don't allow to execute this action on blank/empty file

2. And also this action allows a user to export invalid file, or non-narrative file type and again it throws NullPointerException
[Image from Gyazo](https://gyazo.com/f47d0028c91e96ebf506909fd2e1457c)

### Stacktrace
java.lang.NullPointerException
	at org.mwc.cmap.NarrativeViewer2.NatNarrativeViewer.getVisibleNarratives(NatNarrativeViewer.java:562)
	at org.mwc.cmap.NarrativeViewer2.NATViewerView$19.runWithEvent(NATViewerView.java:833)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$5(ActionContributionItem.java:451)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4131)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1056)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:3944)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3547)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1173)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1062)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:635)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:559)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:155)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:400)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:660)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:597)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1468)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1441)

**Comment by IanMayo** [comment]
Well spotted @helenayele - thanks.

File: org.mwc.cmap.NarrativeViewer/src/org/mwc/cmap/NarrativeViewer2/NATViewerView.java [hunk 1]
```diff
@@ -817,20 +818,25 @@ public void run() {
...
 				}
 				else {
-					dialog.setFilterPath(prefDir);
-				}
-				String targetDirectory = dialog.open();
-				if(targetDirectory!=null) {
-					Activator.getInstance().getPreferenceStore().setValue(PREF_EXPORT_DIR, targetDirectory);
-					NarrativeWriter writer = new NarrativeWriter();
-					writer.write(myViewer.getVisibleNarratives(), myViewer.isShowingSource(), myViewer.isShowingType(), new File(targetDirectory));
+					MessageDialogWithToggle.openWarning(getViewSite().getShell(), "Error exporting narratives", "No narratives to export");
```
Comment (IanMayo) [inline org.mwc.cmap.NarrativeViewer/src/org/mwc/cmap/NarrativeViewer2/NATViewerView.java:839]: @ayeshare - is this the correct dialog class?  We don't have anything to toggle.

**Review by helenayele — APPROVED** [review]
[Image from Gyazo](https://gyazo.com/e19de0063c151b8611c023b667218560)
able to get this dialog when narrative viewer does not have any narrative entries.
so approving...

**Comment by IanMayo** [comment]
@ayeshare - just the one potential change, please: https://github.com/debrief/debrief/pull/4948#discussion_r476412243

File: org.mwc.cmap.NarrativeViewer/src/org/mwc/cmap/NarrativeViewer2/NATViewerView.java (hunk 1 above)
Comment (ayeshare) [reply org.mwc.cmap.NarrativeViewer/src/org/mwc/cmap/NarrativeViewer2/NATViewerView.java]: @IanMayo thats right. Changed it to MessageDialog.

**Review by IanMayo — APPROVED** [review]
Thanks @ayeshare

**Comment by IanMayo** [comment]
@ayeshare - I know it would make the logic a little more complex, but could you display the columns in MS-Word in the same order as in Debrief.

So: Date, Source (optional), Type (optional), Entry.

I suggest you use an incrementing counter for the column numbers.

**Review by helenayele — CHANGES_REQUESTED** [review]
There is column mismatch in the generated doc and narrative viewer.
See that in Narrative viewer Type is empty but in the generated document Entry is blank.
[Image from Gyazo](https://gyazo.com/c53322b79305e6f0e3c42df1b0acb789)

**Comment by IanMayo** [comment]
@helenayele - @ayeshare is going to be re-implementing some of the logic later on.  I'm sure she'll catch your issue then.

**Comment by helenayele** [comment]
okay Ian, and the dialog now looks like [Image from Gyazo](https://gyazo.com/11c926849252a9fd879036177f340a2d)

**Comment by IanMayo** [comment]
Good testing @ayeshare  :-)

**Comment by IanMayo** [comment]
@helenayele - the export logic has changed. Would you mind re-testing?

**Review by helenayele — CHANGES_REQUESTED** [review]
There is a problem in `show source` and `show type' options.
See the below video here.
[Image from Gyazo](https://gyazo.com/ca1700bf88bd5aab64640a56f013d3df)
Show source and show type options are unchecked but they are shown in narrative viewer. This happens as follows:
1. Open sample.dpf
2. Open narrative viewer
3. See that narrative viewers are shown with all columns shown
4. Uncheck `Show source` and `Show type` so that Source and Type columns will be hidden and it looks like [Image from Gyazo](https://gyazo.com/2edd77af67e0c8c48def17d6e0431368)
5. Go to in outline view and uncheck one of the narrative entries there
6. It shows both `source` and `Type` columns but if you check `show source` and `show type` they remain unchecked
[Image from Gyazo](https://gyazo.com/78cd9a19e502de5633126faabbcb6286)
7. And if you try to export to word doc , it doesn't show the source and type columns.
[Image from Gyazo](https://gyazo.com/bd95c8a92edf338dcc8606c0b3c1a5fc). This is because `show source` and `show type` options are unchecked.

**Comment by IanMayo** [comment]
Ok. @helenayele - I think that's unrelated to this PR.  Could you capture it as a separate issue please?

Any other issues with export to MS Word @helenayele ?

**Comment by helenayele** [comment]
> Ok. @helenayele - I think that's unrelated to this PR. Could you capture it as a separate issue please?
Okay filed here #4959, but this issue seems critical to me.
> Any other issues with export to MS Word @helenayele ?
No more issues I observed so far.

**Comment by IanMayo** [comment]
> Okay filed here #4959, but this issue seems critical to me.

Yes, it's important. But the user(s) that use "Export to Word" don't hide/reveal columns in the Narrative Viewer.  So, we can deliver the capability to them without delaying for that fix.

**Pull request merged**